beacon-chain/sync/pending_blocks_queue.go

```diff
@@ -132,85 +116,123 @@ func (s *Service) processPendingBlocks(ctx context.Context) error {
 				continue
 			}
 
+			// Request parent block if not in the pending queue and not in the database.
 			parentRoot := b.Block().ParentRoot()
-			parentInDb := s.cfg.beaconDB.HasBlock(ctx, parentRoot)
-			hasPeer := len(pids) != 0
-
-			// Only request for missing parent block if it's not in beaconDB, not in pending cache
-			// and has peer in the peer list.
-			if !inPendingQueue && !parentInDb && hasPeer {
-				log.WithFields(logrus.Fields{
-					"currentSlot": b.Block().Slot(),
-					"parentRoot":  hex.EncodeToString(bytesutil.Trunc(parentRoot[:])),
-				}).Debug("Requesting parent block")
-				parentRoots = append(parentRoots, b.Block().ParentRoot())
-
-				span.End()
+			inPendingQueue := s.isBlockInQueue(b.Block().ParentRoot())
+			if s.shouldRequestParentBlock(ctx, inPendingQueue, parentRoot) {
```

Comment:
Rather than set `inPendingQueue` here and pass as a bool to `shouldRequestParentBlock`, I would change the impl of `shouldRequestParentBlock` to:
```
func (s *Service) shouldRequestParentBlock(ctx context.Context, inPendingQueue bool, parentRoot [32]byte) bool {
	return s.isBlockInQueue(b.Block().ParentRoot()) && !s.cfg.beaconDB.HasBlock(ctx, parentRoot) && len(s.cfg.p2p.Peers().Connected()) > 0
}
```

Reply:
Actually, I would re-order that boolean evaluation to make the best use of short-circuit evaluation, cheapest to most expensive:
```
func (s *Service) shouldRequestParentBlock(ctx context.Context, inPendingQueue bool, parentRoot [32]byte) bool {
	return len(s.cfg.p2p.Peers().Connected()) > 0 && s.isBlockInQueue(parentRoot) && !s.cfg.beaconDB.HasBlock(ctx, parentRoot) && 
}

```

Reply:
But we should still pass `inPendingQueue` because that's determined before `checkIfBlockIsBad`. Safer that way imo

Reply:
yeah that comment is outdated given the other considerations. Now I would argue we should just make a helper for the peer count and ditch this one - the func is basically converging on comparing 3 booleans.